source/differentiation/ad/ad_helpers.cc

```diff
@@ -1032,6 +1032,146 @@ namespace Differentiation
     }
 
 
+    /* ------------------- ADHelperResidualLinearisation ------------------- */
+
+
+
+    template <enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
+    ADHelperResidualLinearisation<ADNumberTypeCode, ScalarType>::
+      ADHelperResidualLinearisation(const unsigned int n_independent_variables,
+                                    const unsigned int n_dependent_variables)
+      : ADHelperCellLevelBase<ADNumberTypeCode, ScalarType>(
+          n_independent_variables,
+          n_dependent_variables)
+    {}
+
+
+
+    template <enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
+    void
+    ADHelperResidualLinearisation<ADNumberTypeCode, ScalarType>::
+      register_residual_vector(const std::vector<ad_type> &residual)
+    {
+      Assert(residual.size() == this->n_dependent_variables(),
+             ExcMessage(
+               "Vector size does not match number of dependent variables"));
+      for (unsigned int i = 0; i < this->n_dependent_variables(); ++i)
+        ADHelperBase<ADNumberTypeCode, ScalarType>::register_dependent_variable(
+          i, residual[i]);
+    }
+
+
+
+    template <enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
+    void
+    ADHelperResidualLinearisation<ADNumberTypeCode, ScalarType>::
+      compute_residual(Vector<scalar_type> &values) const
+    {
+      if (this->keep_values == false ||
+          ADNumberTraits<ad_type>::is_tapeless == true)
+        {
+          Assert(
+            this->n_registered_independent_variables() ==
+              this->n_independent_variables(),
+            ExcMessage(
+              "Not all values of sensitivities have been registered or subsequently set!"));
+        }
+      Assert(this->n_registered_dependent_variables() ==
+               this->n_dependent_variables(),
+             ExcMessage("Not all dependent variables have been registered."));
+
+      // We can neglect correctly initializing the entries as
+      // we'll be overwriting them immediately.
+      if (values.size() != this->n_dependent_variables())
+        values.reinit(this->n_dependent_variables(),
+                      true /*omit_zeroing_entries*/);
```

Comment:
Where does this happen? In user code or in the call to `values()` below? If the former, this seems like a dangerous optimization maybe not worth the risk.

Reply:
No, its the latter, i.e. in the call to `values()` below. So I think that this is fine (all my tests and production code seem to agree).

Reply:
I've now improved the in-code comment to better clarify this point.